stdlib/public/runtime/MetadataLookup.cpp

```diff
@@ -243,10 +247,112 @@ swift::_matchMetadataByMangledTypeName(const llvm::StringRef typeName,
   return metadata;
 }
 
-// returns the type metadata for the type named by typeName
+// returns the generic type metadata for type named by boundGenericNode
+static const Metadata *
+matchMetadataGenericNode(const llvm::StringRef typeNodeName,
+                         const NodePointer boundGenericNode,
+                         const Metadata *unused __attribute__((unused)),
+                         const NominalTypeDescriptor *ntd) {
+  if (boundGenericNode->getNumChildren() != 2)
+    return nullptr;
+
+  auto templateTypeNode = boundGenericNode->getChild(0);
+  auto typeListNode = boundGenericNode->getChild(1);
+
+  if (templateTypeNode->getKind() != Node::Kind::Type ||
+      typeListNode->getKind() != Node::Kind::TypeList)
+    return nullptr;
+
+  // Validate that we have one or more generic parameters (the zero parameter
+  // case is handled by matchMetadataTypeNode) and that the number of primary
+  // archetypes matches the number of child nodes in the demangled type.
+  //
+  // TODO: support secondary archetypes, where NumParams != NumPrimaryParams;
+  // not sure if there is enough information available at runtime to do this.
+  auto &genericParams = ntd->GenericParams;
+  if (!genericParams.hasGenericParams() ||
+      genericParams.NumPrimaryParams != typeListNode->getNumChildren() ||
+      genericParams.NumParams != genericParams.NumPrimaryParams)
+    return nullptr;
+
+  auto templateName = mangleNode(templateTypeNode); // XXX cache this
+  if (ntd->Name.get() != templateName)
+    return nullptr;
+
+  auto pattern = ntd->getGenericMetadataPattern();
+  void **arguments = new void *[pattern->NumKeyArguments];
+
+  for (size_t i = 0, witnessIndex = 0;
+       i < genericParams.NumPrimaryParams; i++) {
+    auto param = genericParams.getParameterAt(i);
+    auto metadata = findMetadataForNode(typeListNode->getChild(i));
+    if (metadata == nullptr) {
+      // unknown type
+      delete arguments;
+      return nullptr;
+    }
+
+    for (size_t j = 0; j < param->NumWitnessTables; j++, witnessIndex++) {
```

Comment:
I see you're recording the generic constrains in the form of references to protocols from the NominalTypeDescriptor.

One thing to investigate might be to mangle the generic signature and store that in the NTD as a string instead. Then you will be able to capture other types of constraints, including constraints on associated types, and as long as you can look up protocols by name also, you will be able to implement the same functionality.